### PR TITLE
CanIUse: Adding notes to abstract as well as related links

### LINKED ISSUE #245: Improve CanIUse Fathead
### Objective

Improve the display of the results, and triggering for the CanIUse Fathead

### Background

CanIUse currently addresses most of the CSS queries (1.7%) but the Instant Answer also addressed various HTML and JavaScript queries.

Visually, it's not the most appealing Instant Answer, and it also has room for triggering improvements.

### ToDo

1. Queries containing "compatibility" don't trigger this Instant Answer and they should:

 - [css3 colors compatibility](https://duckduckgo.com/?q=css3+colors+compatibility&ia=web)
 - [flexbox css browser compatibility](https://duckduckgo.com/?q=flexbox+css+browser+compatibility&ia=web)

2. Add the **notes** to the result, or at least provide a link to them
3. Create an InfoBox with external links to the **resources** to provide quick access to more information

### What to do: 

1. Comment here if you'd like to handle this task. 
2. Review any Related Discussions, Issues, and Pull Requests (below).
3. [Setup your Dev Environment](http://docs.duckduckhack.com/welcome/setup-dev-environment.html) (first-timers only)
4. Refer to the [Fathead documentation](http://docs.duckduckhack.com/resources/fathead-overview.html) and [other Fatheads](https://duck.co/ia?repo=fathead) as examples
5. [Submit a Pull Request](http://docs.duckduckhack.com/submitting/pull-request.html) with your completed changes

### Who to contact for help: 
Please contact @moollaza  or see [other contact info](http://docs.duckduckhack.com/resources/get-in-touch.html).

----

**Related Discussions, Issues, and Pull Requests:** 

Forum Topic: https://forum.duckduckhack.com/t/improve-ia-caniuse-fathead/228/1

### PR DESCRIPTION
**Pull request by samskeller**
Fixes #245. I've added the 'notes' from the input data to the 'abstract' field as well as adding the links from the input data to the 'related' field in the IA. Also added support for the "TP" release version used by Safari to indicate "Technology Preview" -- we treat this as the latest version of the browser. 

This will requires some testing to make sure the links show up right as they might not be at the moment. It looks like to fix the triggering issue we should add some internal triggering words (specifically "compatibility") to fix everything discussed in issue 245. 

---

https://duck.co/ia/view/caniuse
Maintainer: @csytan

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by edgesince84** [comment]
thanks for the PR!  will review shortly - cc @moollaza

File: lib/fathead/caniuse/parse.py [hunk 1]
```diff
@@ -31,15 +31,25 @@ def generate_answers(data):
...
-                prefix=agent['prefix'], 
+                browser=agent['browser'],
+                prefix=agent['prefix'],
                 stats=feature_data['stats'][browser])
             abstract += '<br>' + out
+
+        # Add notes to abstract, if there are any
+        notes = feature_data.get('notes', '')
+        if notes:
+            abstract += '<br>' + notes
```
Comment (moollaza) [inline lib/fathead/caniuse/parse.py]: @samskeller I'd like to make the notes a little more obvious and separate them from the other content.

Let's wrap the notes in their own `<p>` which will space things out a bit. We should also add "Notes:" above to make it clear. The final output should look like:

```html
<p><b>Notes:</b> $notes</p>
```

Also, it looks like the notes (and now also the descriptions) contain Markdown. We should either convert the markdown to HTML, or strip out the HTML.

Would you like to try converting the Markdown to HTML first? I believe that approach will work.

File: lib/fathead/caniuse/parse.py [hunk 2]
```diff
@@ -31,15 +31,25 @@ def generate_answers(data):
         ])
         print ','.join(titles)
 
+        # Generate Related Topics
+        links = feature_data.get('links', [])
+        related = u' '.join([u'[[{} {}]]'.format(link['title'], link['url']) for link in links])
+
         # Generate abstract
         abstract = feature_data['description']
```
Comment (moollaza) [inline lib/fathead/caniuse/parse.py]: @samskeller let's also wrap the text description in its own `<p></p>` to separate it from the list of supported browsers. Right now it's one giant text blob that's hard to understand:

flexible_box_layout_module__browser_support__at_duckduckgo

File: lib/fathead/caniuse/parse.py [hunk 3]
```diff
@@ -31,15 +31,25 @@ def generate_answers(data):
         ])
         print ','.join(titles)
 
+        # Generate Related Topics
+        links = feature_data.get('links', [])
+        related = u' '.join([u'[[{} {}]]'.format(link['title'], link['url']) for link in links])
```
Comment (moollaza) [inline lib/fathead/caniuse/parse.py]: Unfortunately I was wrong about the way Related Topics work. They can only be internal links to other Fathead articles so this isn't going to work.

Let's remove or comment this out for now 👍

**Comment by samskeller** [comment]
@moollaza thanks for the feedback -- I've updated my code and tried to update my fork with the changes from upstream but I appear to have done something wrong as there's a bunch of other changed files in this PR now... any chance you see a way to fix what I did? I think I just did a `git rebase upstream/master` followed by a `git pull` and `git push`...

**Comment by moollaza** [comment]
@samskeller unfortunately rebasing has this effect, it's usually safer to `git pull` and resolve any conflicts (though there shouldn't have been any).

Technically this OK, though it does look a bit messy.

You could `git reset --hard` your branch to reset it, and then `git cherry-pick` your individual commits, then `git push -f` to reset the branch?

**Comment by samskeller** [comment]
@moollaza great fix! That appears to have worked. This is ready for review again

**Comment by moollaza** [comment]
@samskeller this is now live on Beta!

https://beta.duckduckgo.com/?q=caniuse+CSS3+2D+Transforms&t=hk&ia=about&iax=1

I noticed https://beta.duckduckgo.com/?q=caniuse+calc()+as+CSS+unit+value&t=hk&ia=about&iax=1 isn't working correctly.

We may actually need to convert the notes from markdown, to plaintext instead of HTML. I will confirm if Fatheads strip out other HTML like we have here..

**Comment by moollaza** [comment]
@samskeller I've confirmed that `<a>` tags in the Fathead abstract have been causing issues. If we strip out the `<a>` tag, the abstract should render as expected 👍

**Comment by samskeller** [comment]
@moollaza okay cool -- so I should just strip out `<a>` tags but leave all other HTML tags?

**Comment by moollaza** [comment]
> @moollaza okay cool -- so I should just strip out <a> tags but leave all other HTML tags?

@samskeller yes! Let's try that. Other HTML should be fine 👍

**Comment by samskeller** [comment]
@moollaza I've added the `BeautifulSoup` library as a requirement for this Fathead as I used it to remove the `<a>` tags -- so this should be ready for review again!

**Comment by moollaza** [comment]
@samskeller it looks like the abstracts are working correctly! Deploying to Beta right now...

File: lib/fathead/caniuse/parse.py [hunk 4]
```diff
@@ -31,15 +33,33 @@ def generate_answers(data):
...
             abstract += '<br>' + out
+
+        # Add notes to abstract, if there are any
+        notes = feature_data.get('notes', '')
+        if notes:
+            notes_in_html = markdown.markdown(notes)
+            bs = BeautifulSoup(notes_in_html)
+            for a in bs.findAll('a'):
+                a.replaceWithChildren()
+            abstract += '<p><b>Notes:</b> {}</p>'.format(bs.renderContents())
```
Comment (moollaza) [inline lib/fathead/caniuse/parse.py]: Let's put `<b>Notes:<b><br>` inside the `<p>` that contains the `description`

That way we'll separate the "Notes:" text from the text above.

We should also wrap the browser support lines in one `<p>`.

The output should look like:

```html
<div class="c-info__content chomp js-ellipsis">
  
  <p>Method of positioning elements in horizontal or vertical stacks. Support includes the support for the all properties prefixed with `flex` as well as `display: flex`, `display: inline-flex`, `align-content`, `align-items`, `align-self`, `justify-content` and `order`.</p>
  
  <p>IE 11+ partial<br>Firefox 28+ ✓
    <br>Chrome 29+ ✓
    <br>Safari 9+ ✓
    <br>iOS Safari 9.0-9.2+ ✓
    <br>Android Browser 4.4.3-4.4.4+ ✓
  </p>

  <p>
    <b>Notes:</b>
    <br>Most partial support refers to supporting an older version of the specification or an older syntax.
  </p>

</div>
```

This way the output looks like:

flexible_box_layout_module_at_duckduckgo

**Comment by moollaza** [comment]
@samskeller this is looking pretty good. After a bit more thought, I think we have a lot of potential with the design of this Fathead, I'm thinking we should move to a table for the browser support and we can have it look something like:

flexible_box_layout_module_at_duckduckgo

I'm going to merge this for now and open a new issue for that 👍

**Comment by russellholt** [comment]
testing some data automation issues with this comment

**Pull request merged**